### ISSUE TITLE
Option for http.errors to handle errors thrown from php (fastcgi)

### ISSUE THREAD

**Issue by infinityCounter** [mholt/caddy#1882]
### 1. What version of Caddy are you using (`caddy -version`)?
0.10.9

### 2. What are you trying to do?
Log error responses (HTTP Status Code >=400) using the errors middleware

### 3. What is your entire Caddyfile?
```
localhost { 
  tls self_signed
  root ./App/
  gzip
  log / ./App/Logs/caddyaccess.log "{remote} - {user} [{when}] \"{method} {uri} -> {rewrite_path} {proto}\" {status} {size}" {
  	  rotate_size 50  # Rotate after 50 MB
  	  rotate_age  90  # Keep rotated files for 90 days
  	  rotate_keep 20  # Keep at most 20 log files
  	  rotate_compress # Compress rotated log files in gzip format
  }
  errors ./App/Logs/caddyerrors.log {
	  rotate_size 50 
	  rotate_age  90 
	  rotate_keep 20
	  rotate_compress 
  }

  # Following rewrite blocks does an internal
  # URL rewrite for all resources (NOT PHP FILES) to ./App/Assets folder
  rewrite / {
 	  regexp /.*\..+
      ext    !.php
      to     /Assets/{path}
  }

  # Rewrite all requests for anything else to root
  rewrite / {
   	  to     /
  }

  fastcgi / /var/run/php/php7.0-fpm.sock {
      ext .php
      split .php
      index index.php
      connect_timeout "120s"
  }
}
```

### 4. How did you run Caddy (give the full command and describe the execution environment)?
caddy (Ran from born again shell in directory with Caddyfile)

### 5. Please paste any relevant HTTP request(s) here.

Not relevant

### 6. What did you expect to see?
I expected the content of my 500 response to be logged to into ./App/Log/caddyerrors.log

### 7. What did you see instead (give full error messages and/or log)?
No 500 response were logged

### 8. How can someone who is starting from scratch reproduce the bug as minimally as possible?
Start a php application and try the following code in an index.php and test from any browser
```
try{
    throw new Exception("Problemo");
}catch(Exception $e)
{
    http_response_code(500);
    echo $e->getMessage() . ", this should be in your log file but it's not.";
    flush();
}
```

**Comment by tobya** [comment]
Thank you for the issue.  I assume by the way you have phrased your question that if you set the reponse code to < 400 it does get printed?  Does anything ever get printed to the error log?

I'm wondering does the server consider a status code actually an error?

404 are logged to the Log but not to the error file, as it is not an 'error' in that sense.

500 if anything I think probably get logged to the main webserver log set by the -log directive on startup.

Let me know the answer to the first question.

**Comment by mholt** [comment]
I haven't looked at the fastcgi code in a while, but I think if the backend writes a response, then Caddy does not; or in other words, the backend is expected to handle errors. But I'm not sure on that. Would be worth verifying when we have a chance.

**Comment by infinityCounter** [comment]
@tobya If I set it to < 400 it does not print to the error log, nothing is ever printed to the error log as long as my code sends a response. In the case where there's a fall through and my code does not send a response, then caddy logs an error.

@mholt  you're write about the fastcgi code, I just took a browse through and responses aren't checked for their status code by the fastcgi. It assumes all responses are Ok and shouldn't be considered an error.
